flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -288,9 +288,7 @@ BufferAndBacklog pollBuffer() {
 
             if (buffers.isEmpty()) {
                 flushRequested = false;
-            }
-
-            while (!buffers.isEmpty()) {
+            } else {
```

Comment:
After the changes, if the first buffer is finished and empty. The empty buffer will be sent to the downstream to release the allocated credit for it instead of recycled directly.

Reply:
If there is only one buffer in the buffer queue, after all data is read, the flushRequested flag will be turned off. Consider the following scenario (correct me if I am wrong):

1. There is an unfinished buffer in the queue and all data is read.
2. Then the buffer is finished but no new data is appended. Note appending data and finish buffer is not an atomic operation.
3. New buffer or event is added to the buffer queue.
4. When polling buffer, an empty buffer is at the head of the queue and we already allocate a credit for it.

Reply:
> appending data and finish buffer is not an atomic operation.
It is exactly what I want to emphasize. As I understand, we need the empty buffer only because `appending data and finish buffer is not an atomic operation.` which happens because BufferBuilder has a gap between `commit` and `finish`. So I propose to rewrite BufferBuilder in such a way that if it commits the current state and size == capacity it finishes it immediately(so there is no explicit finish call needed). These changes allow us to avoid all logic with an empty buffer which I think pretty untransparent.

One more time, my proposal:

- Rewriting the `BufferBuilder` in such a way that it will be impossible to have the empty finished buffer(it just requires removing the gap between the last commit and finish).
- Don't send the empty buffer(after changes described above there won't be any empty buffers)

@wsry WDYT?
@pnowojski What is your opinion, is it safe to finish BufferBuilder immediately when it becomes full? I mean not to wait for call explicit `finish` but finish in the `commit` if it is full?

Reply:
Does that mean that the BufferBuilder can not finish a partial buffer, for example the last piece of data of the batch execution mode? Besides, there is another case which may lead to the empty buffer: in PipelinedApproximateSubpartition, if a partial record takes more than one buffer in the queue, those partial data will be dropped when recovering.

Reply:
Yes, the batch execution mode still requires the manual calling of finish which brokes my idea.  The same about PipelinedApproximateSubpartition. As I understand, it is impossible to know that a partial record will be sent or not before the request the credit.
So I still believe that it would be better to avoid sending the empty buffer if it is possible. But perhaps you right and it will be not so easy to do so in the current implementation, especially because of the scenarios which you described.

Reply:
Can we measure how many those empty buffers are being sent? Is this an issue of 0.1% more messages or 50%? If it's hard to say or can vary, would it be difficult/would it complicate the code to send the empty buffers only if `#exclusiveBuffers == 0` to make sure that there is no performance regression?

Reply:
It is pretty simple to send empty buffer only when  #exclusiveBuffers == 0.